src/utils/batchDiscordRequests.ts

```diff
@@ -0,0 +1,133 @@
+import JSONResponse from "./JsonResponse";
+import { addDelay, convertSecondsToMillis } from "./timeUtils";
+export const DISCORD_HEADERS = {
+  RATE_LIMIT_RESET_AFTER: "X-RateLimit-Reset-After",
+  RATE_LIMIT_REMAINING: "X-RateLimit-Remaining",
+  RETRY_AFTER: "Retry-After",
+};
+
+const MAX_RETRY = 1;
+const LIMIT_BUFFER = 0.2;
+
+interface RequestDetails {
+  retries: number;
+  request: () => Promise<Response>;
+  index: number;
+}
+interface ResponseDetails {
+  response: Response;
+  data: RequestDetails;
+}
+
+const parseRateLimitRemaining = (response: Response) => {
+  let rateLimitRemaining = Number.parseInt(
+    response.headers.get(DISCORD_HEADERS.RATE_LIMIT_REMAINING) || "0"
+  );
+  rateLimitRemaining = Math.floor(rateLimitRemaining * (1 - LIMIT_BUFFER));
+  return rateLimitRemaining;
+};
+
+const parseResetAfter = (response: Response) => {
```

Comment:
bugs in the code:
Loss of precision: Parsing Rate-Limit-Reset-After header using parseFloat might not be ideal as it can lead to loss of precision for non-integer values. This could be especially problematic if the header contains fractional seconds.

Missing overflow handling: Similar to the previous code, the parsed value is multiplied by 1 and then rounded up using Math.ceil. This approach could potentially lead to overflows for large values, causing incorrect calculations.

Reply:
>  Parsing Rate-Limit-Reset-After header using parseFloat might not be ideal as it can lead to loss of precision for non-integer values. 

Here we are expecting seconds within the range of 0 - 60. `parseFloat` does support decimal values.

> Missing overflow handling: Similar to the previous code, the parsed value is multiplied by 1 and then rounded up using Math.ceil

- Here we are not multiplying.
- Math.ceil is to be on a safer side and avoid getting rate limit by milliseconds. 
- Since we are not expecting a huge number for seconds, I don't see any reason to use other methods which can handle large number like bigInt